docs/model-deployment-prepackaged-server-custom.md

```diff
@@ -13,15 +13,159 @@ description: Custom Pre-packaged Server
+        if self.ready:
+            return
+
+        # build model
+        # 1. set to self.model
+        # 2. make set.ready = True
+        self.ready = True
+
+    def predict(self, X, feature_names = None, meta = None):
+        # execute self.model.predict(X)
+        print(X, feature_names, meta)
+        return "Hello Model"
+```
+
+
+## Handle model files
+
+The model files could ship with your image, you could make it more feasible with the `model_uri` mechanism of the PrimeHub Deployment.
+
+PrimeHub users could create a Deployment with Model URI:
+![](assets/mdeploy_quickstart_deploydetail_1_phfs.png)
+
+In the container's view, the model files will mount to a local filesystem, such as `/mnt/models`:
+
+```python
+def __init__(self, model_uri=None):
+    self.model_uri = model_uri
+    ...
+```
+
+Please add the `model_uri` argument to the `__init__` function and make sure having `None` as the default value. It means your Model supporting with or without `model_uri`. If a user give a `Model URI` value, the `__init__` will get a mount path from the variable `model_uri`. You can check which files should be loaded in that path.
+
+It is very common to write a load method to load and build model instance:
+
+```python
+  def load(self):
+    ...
+```
+
+You could build the model instance in the `__init__`, if your loading process is very simple. The `load` method is optional.
+
+## MLflow model
+
+PrimeHub Deployment could work with MLflow Model Registry. When a user set `Model URI` starting with `models:`, it will get model files from MLflow Model Registry. Nothing special with the model files comes from the MLflow Model Registry, it also mounts to the `model_uri`. However, there is a convention to let you know:
+
+> The model files always ships with the `MLmodel` file.
+
+You should register a model with the `MLmodel` file
+
+### Register model
+
+Here is an example to train and log model to the MLflow tracking server:
+
+```python
+import mlflow
+import numpy as np
+from sklearn.linear_model import LinearRegression
+
+# it will set from env MLFLOW_TRACKING_URI
+# mlflow.set_tracking_uri("...")
+mlflow.set_experiment("experiment-1")
+
+X = np.array([[1, 1], [1, 2], [2, 2], [2, 3]])
+y = np.dot(X, np.array([1, 2])) + 3
+
+# train a model
+model = LinearRegression()
+with mlflow.start_run() as run:
+    model.fit(X, y)
+    print("Logged data and model in run {}".format(run.info.run_id))
+    mlflow.sklearn.log_model(model, "models/model-a")
+```
+
+After the model uploaded, it could be found in the **experiment-1**:
+
+![](assets/mlflow-register-model-with-MLmodel.png)
+
+When you select a directory in the artifact browser, it will show a `Register Model` botton in the right side, please register the directory which contains the `MLmodel` file. Let's check it one by one:
+
+1. (invalid) **models** directory: it doesn't contain a MLmodel file
+1. (valid) **model-a** directory: it contains a MLmodel file
```

Comment:
Actually, we don't need to have this restriction. It is still ok if there is no MLmodel inside. As long as the custom pre-packaged server can consume the content in the model.

Reply:
got it :D